src/metabase/models/permissions.clj

```diff
@@ -229,19 +243,36 @@
    (table-segmented-query-path (:db_id table) (:schema table) table))
 
   ([database-or-id schema-name table-or-id]
-   (str (object-path (u/get-id database-or-id) schema-name (u/get-id table-or-id)) "query/segmented/")))
+   (str (object-path (u/the-id database-or-id) schema-name (u/the-id table-or-id)) "query/segmented/")))
 
 
 ;;; -------------------------------------------- Permissions Checking Fns --------------------------------------------
 
+(defn- moderate?
+  [path]
+  (str/ends-with? path "/moderate/"))
+
+(defn- path-without-access-level
+  [path]
+  (str/replace path #"(read|moderate)/$" ""))
+
 (defn is-permissions-for-object?
   "Does `permissions-path` grant *full* access for `object-path`?"
   [permissions-path object-path]
-  (str/starts-with? object-path permissions-path))
+  (if (= "/" permissions-path)
+    true
+    (if (moderate? permissions-path)
+      ;; Moderate permission has access to everything in that path
+      (str/starts-with? object-path (path-without-access-level permissions-path))
+      (and (not (moderate? object-path)) ;; We don't have moderate permission
+           ;; Hack since /collection/1 permission grants access to both /collection/1 and /collection/1/read
+           (str/starts-with? object-path permissions-path)))))
```

Comment:
This is pretty futzy code and probably the most important function in the entire PR. I'm sure a student of predicate logic and string hacks could tighten it up a bit, but I thought being explicit about some of the different cases was helpful

Reply:
Definitely not a big fan of doing special-case hacks to the permissions system. The whole house of cards is going to come crashing down if we invert how things work, since other things rely on the prefix-based nature of the perms system (such as DB queries that do `LIKE /something/%`. **Please** don't add this stuff 😬

Reply:
I don't think the whole concept of "Moderate permission has access to everything in that path" makes any sense.

There are two ways you can modify a Collection: 

-  "Editing" it by adding or removing items, changing the name, etc.
- Add reviews

Both of these are subsets of "write" perms for a Collection. So they should both go under a common prefix, and then you can give people one or the other or just the entire prefix if you want them to have both. For example you could do something like

```
/collection/1/        -> full perms to do anything for Collection 1
/collection/1/edit/   -> perms to edit collection and add or remove items to Collection 1
/collection/1/review/ -> perms to add/remove reviews for Collection 1
```

Now assuming you want the permission level to be are none < read-only < "Edit" < "Moderate" you should grant permissions as follows:

```
Read perms     -> /collection/1/read (no perms to /edit/ or /review/
Edit perms     -> /collection/1/edit/ and /collection/1/read/ (perms to read, and to /edit/)
Moderate perms -> /collection/1/ (full perms to do anything to Collection 1, including read, edit, and review)
```

Then you can go around and check whether people have `/collection/1/review/` perms, which is true whether they're an admin (they have `/` perms) or someone else you granted moderate perms to (they have `/collection/:id/` perms).

You'll also want to write a migration to migrate existing `/collection/:id/` perms in the DB to `/collection/:id/edit/` since the meaning has changed. This should be easy to do tho. You should definitely do it as a Liquibase migration instead of a Clojure-land one because the Clojure-land stuff is so fragile

Reply:
Thinking about this a bit more.

You might consider tweaking the paths a bit so you don't need to grant two entries to give someone Read + "Edit" perms. Read perms seem like a subset of whatever "Edit" perms are.

That was previously reflected by having `/collection/1/` be the common prefix -- `/collection/1/` meant  "readwrite" and `/collection/1/read/` meant "read", a subset of "readwrite". Maybe the new paths should look like

```
/collection/1/readwrite/read/ -> perms to Read 
/collection/1/readwrite/      -> perms to Edit, and to Read
/collection/1/review/         -> perms to add/remove reviews 
/collection/1/                -> perms to do anything, including Edit, Read, or "Moderate"
```

I'm cool with whatever approach as long as we don't flip the whole prefix concept on its head

Reply:
Another alternative to consider:

for a few models we consider you to have "read" perms if you have "partial perms" for an object, e.g. any perms starting with that object's prefix. e.g.

```clj
(can-read? Collection 1) 
-> 
(user-has-partial-perms? "/collection/1/") 
-> 
user has any perms starting with /collection/1/
```

You could set it up this way and then stick with `/collection/1/read/`, `/collection/1/edit/` and `/collection/1/review/` and not have to worry about migrating `/read/` to `/readwrite/read`, or granting read + edit perms at the same time.

I'm leaning towards this approach out of the 3 I've suggested since it would require the least amount of changes. I'm cool with any approach that keeps the spirit of the perms system intact tho.

Reply:
Thanks for the thoughtful feedback. To synthesize, it sounds like the easiest fix is:

* Keep the new `/collection/:id/moderate/` as-is. Moderation necessarily implies readwrite access.
* Keep the old `/collection/:id/read/` as-is
* Introduce `/collection/:id/edit/` to mark readwrite (but not moderation) permission.
* `/collection/:id/` will be the "do-everything" permission, but in practice it just means that admins who have `/` will be able to do what they want.

However, this doesn't solve the problem of needing permissions to be a strict prefix of what they grant access too. E.g., a permission of `LIKE '/collection/42/moderate/%'` won't work (but should) if the item path is `/collection/42/edit`.

As far as I can figure, the only way to make prefixing work is to stack it. So the paths would be:

| access level | path |
|--|--|
| read collection 42 | `/collection/42/moderate/edit/read/`
| edit collection 42 (implies read access) | `/collection/42/moderate/edit/`
| moderate collection 42 (implies read+edit access) | `/collection/42/moderate/`

This is intuitively a little gross to me, and more specifically I think it's brittle (it'll be a PITA if we add a new type of permission) and I would prefer not to write a DB migration to change every single path in the DB (i.e., `collection/42/read/` → `collection/42/moderate/edit/read/`)

If there's any way to drop the prefix requirement, I think everything would be cleaner. I would have thought that permission-checking should happen at the application layer (or at least with carefully-tuned functions that spit out the right SQL) and hope we don't have ad hoc SQL being written all over the place. IMHO good encapsulation is to only deal with the exact contents of the `permission` table in `models/permission.clj` and let the rest of the codebase deal with the `can-read?` / `can-edit?` / `can-moderate?` functions.

Reply:
I think our wires are a bit crossed. Let me try to explain what I was proposing a little more clearly.

For perms that you'd _grant_, you'd grant:

| access level | path |
|--|--|
| Read-only access to Collection 1 | `/collection/1/read/` |
| "Edit" Collection 1 | `/collection/1/edit/` |
| "Full" perms for Collection 1 including edit or 'moderate | `/collection/1/` |

For perms that you'd _check_, you'd check

| Perms to Check | How to Check | (SQL equivalent |
|--|--|--|
| Current user can read? | current user has partial (any) perms? | `perms LIKE '/collection/1/%'` |
| Current user can edit? | current user has edit perms? | `concat(perms, '%') LIKE '/collection/1/edit/'` |
| Current user can 'review (moderate?)' | current user has 'review' (?) perms | `concat(perms, '%') LIKE '/collection/1/review/'` |

So there's no special cases where presence of any particular permission implies any other. We already have tooling for the "partial perms" stuff (and use it in several other places) as well as code for generating the sorts of SQL in the example above. So you could do all of this without changing any existing tooling. The only thing you'd have to change is migrating existing perms like `/collection/1/` to `/collection/1/edit/`, but you can do that with a simple SQL statement in Liquibase

Reply:
Makes sense, many thanks!